### PR TITLE
(GH-99) Very simple support for Code Outline

### PR DESCRIPTION
**Pull request by Tdue21**
First go at Code Outline functionality.
It just populates the Code Outline View with the list of Tasks in the current file. 
Nothing fancy. 

I am certainly someone with a greater understanding of TypeScript and VSCode can improve upon this 😁

### CODE REVIEW

**Comment by gep13** [comment]
@Tdue21 thanks for taking a look at this, I think that this is a really good start.  Other outline items that we might want to look for are Setup/Teardown, as well as TaskSetup/Teardown.

I am wondering whether pre-processor directives is something that we should add to the outline as well, but I am less convinced by those 😄

**Comment by Tdue21** [comment]
Oh, I agree. Both on the regex and the Setup/TearDown. This is the first ever TypeScript code I have made, so I have kept it simple 😄 
Also async support and classes in cake files.

**Comment by gep13** [comment]
@Tdue21 I have just tried to take this for a spin, and I have checked out your code changes.  Is there something that I need to do in order to get the outlining to appear in the window?  I am not seeing anything in there.

I have been discussing this issue with @mholo65 and there is a potential that we can get a lot of the information that we need by calling into Omnisharp to get all the information that we need.  I am hesitant to have you do too much work on this until we figure out whether this is a route that we want to go down.

**Comment by Tdue21** [comment]
@gep13 Hm. I tried to install the vsix without success, no doubt I am just doing something wrong. But it works fine when I debug it, as you can see in this screendump.

image

**Comment by Tdue21** [comment]
Hm. Now I got the vsix from AppVeyor to install and the outline view works fine from that as well.

**Comment by Tdue21** [comment]
Oh, and if you do not wish to go down this path, then that is of course fine. Although a smidgen disappointing since this is my first TypeScript 😉

**Comment by gep13** [comment]
Sorry, perhaps I didn’t explain myself correctly.... I let try again. 😄

We definitely want you to continue with this work, we really appreciate you working on it.  What I am trying to avoid is wasted effort from your part. What you have done so far is great, and sets up the basis for this feature. However, I don’t want us to waste any effort attempting to further parse the Cake file for the different elements if we don’t need to. Omnisharp already fully understands the Cake file, so assuming that we can figure out how to access this information, we can take the information that it returns and add this into the code that you already have, to complete the outlining feature.

Does that make sense?

File: src/documentSymbols/cakeDocumentSymbolProvider.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                    ));
+                }
+            }
+
+            resolve(symbols);
+        });
+    }
+
+    public matchFunction(line: string) {
+        const function_regex = "Task\\s*?\\(\\s*?\"(.*?)\"\\s*?\\)"
```
[the remark this answers is not in the document]
Comment (Tdue21) [reply src/documentSymbols/cakeDocumentSymbolProvider.ts]: Will do.

**Comment by Tdue21** [comment]
Yes, I understand (now). I completely agree as well 😄 
I do intend to dig a little deeper though, although I do not want to implement a whole lot of functionality, if it turns out we can get it from Omnisharp. 

I intend to do the basics though: Detect and list Setup/TearDown, SetupTask/TearDownTask, Tasks and RunTarget. 
By that is more or less just a matter of expanding the reg ex 😄 

I also want to move the regex into the configuration, as you mentioned, and maybe lay the groundwork for more customization.

An issue, by the way, regarding the code outline view, is that there is a fixed number of options for icons in the view, as per this enum defined in vscode.d.ts: 

```
export enum SymbolKind {
		File = 0,
		Module = 1,
		Namespace = 2,
		Package = 3,
		Class = 4,
		Method = 5,
		Property = 6,
		Field = 7,
		Constructor = 8,
		Enum = 9,
		Interface = 10,
		Function = 11,
		Variable = 12,
		Constant = 13,
		String = 14,
		Number = 15,
		Boolean = 16,
		Array = 17,
		Object = 18,
		Key = 19,
		Null = 20,
		EnumMember = 21,
		Struct = 22,
		Event = 23,
		Operator = 24,
		TypeParameter = 25
	}
``` 

Right now I am thinking of using the Constructor value for Setup/TearDown and Function for Tasks. 
Unfortunately I can't find a list of the _icons_, but most of them can be found by opening any substantial ts document 😄

**Comment by Tdue21** [comment]
Unless there is anything you wish me to change, I believe this feature is finished for now.

**Comment by Tdue21** [comment]
Hello, wondering what is holding this back? Are there anything I need to be doing?

**Comment by gep13** [comment]
@Tdue21 no, I don't think that there is anything else that you need to do at the minute.  This is on me to review and get merged in, but my priorities have been elsewhere.  I will get round to this at some point though.

**Comment by Tdue21** [comment]
@gep13 Entirely understandable. I was just wondering 😄

**Comment by nils-a** [comment]
@Tdue21 I realize this has been idling around for some time. Are you in a position and willing to rebase this PR? I'd really love to take it for a spin.

**Comment by Tdue21** [comment]
@nils-a I haven't touched it in almost 2 years. I have honestly completely forgotten about this, until Github notified me of activity. 
I am not in a position right now to look at it, but if someone else could do so, they'd be more than welcome to it :)

**Comment by nils-a** [comment]
@Tdue21 ok, thanks for letting me know. I'll try to get this up and flying again.

**Comment by gep13** [comment]
Closing this in favour of https://github.com/cake-build/cake-vscode/pull/457

**Pull request closed without merging**